swarm/swap/swap.go

```diff
@@ -0,0 +1,347 @@
+		Price:     big.NewInt(100),
+		SizeBased: true,
+		Direction: ChargeReceiver,
+	}
+
+	return priceMatrix
+
+}
+
+//Check if a message needs accounting
+func (dpo *DefaultPriceOracle) IsAccountedMsg(event *p2p.PeerEvent) bool {
+	var protoPriceMap map[uint64]*PriceTag
+	var ok bool
+	if protoPriceMap, ok = dpo.priceMatrix[event.Protocol]; !ok {
+		return false
+	}
+	if _, ok = protoPriceMap[*event.MsgCode]; !ok {
+		return false
+	}
+	return true
+}
+
+//Get the actual price of a message
+//If the message is size-based, it returns the calculated price
+func (dpo *DefaultPriceOracle) GetPriceForMsg(event *p2p.PeerEvent) (*big.Int, EntryDirection) {
+	if dpo.IsAccountedMsg(event) {
+		priceTag := dpo.priceMatrix[event.Protocol][*event.MsgCode]
+		price := &big.Int{}
+		price.Set(priceTag.Price)
+		if priceTag.SizeBased {
+			price.Mul(priceTag.Price, big.NewInt(int64(*event.MsgSize)))
+		}
+		return price, priceTag.Direction
+	}
+	return nil, false
+}
+
+//This swap implementation works by listening to message events on the p2p server.
+//It then handles the event received, filtering for messages and evaluating
+//if it needs accounting
+func (s *Swap) registerForEvents(srv *p2p.Server) {
+	go func() {
+		events := make(chan *p2p.PeerEvent)
+		sub := srv.SubscribeEvents(events)
+		defer sub.Unsubscribe()
+
+		for {
+			select {
+			case event := <-events:
+				go s.handleMsgEvent(event)
+			case err := <-sub.Err():
+				log.Error(err.Error())
+				return
+			}
+		}
+	}()
+}
+
+//Handle the message.
+//Determine if it needs accounting, and if yes, account for it
+func (s *Swap) handleMsgEvent(event *p2p.PeerEvent) {
+	if !s.priceOracle.IsAccountedMsg(event) {
+		return
+	}
+	s.AccountMsgForPeer(event)
+}
+
+//Do the accounting
+//Depending on the charging type of the message (set in the PriceTag),
+//it will charge the sender or receiver
+func (s *Swap) AccountMsgForPeer(event *p2p.PeerEvent) {
+	price, direction := s.priceOracle.GetPriceForMsg(event)
+	if price == nil {
+		//TODO what to do in this case? Should not happen
+		log.Crit("Price is nil; this should have been accounted for but somehow it failed")
+	}
+
+	if direction == ChargeSender {
+		//we are sending a ChargeSender message, thus debit us and credit remote
+		if event.Type == p2p.PeerEventTypeMsgSend {
+			s.chargeLocal(event, price)
+			//we are receiving a ChargeSender message, so credit us and debit remote
+		} else if event.Type == p2p.PeerEventTypeMsgRecv {
+			s.chargeRemote(event, price)
+		}
+	} else {
+		//we are receiving a ChargeReceiver message, thus debit us and credit remote
+		if event.Type == p2p.PeerEventTypeMsgRecv {
+			s.chargeLocal(event, price)
+			//we are sending a ChargeReceiver message, thus credit us and debit remote
+		} else if event.Type == p2p.PeerEventTypeMsgSend {
+			s.chargeRemote(event, price)
+		}
+	}
+}
+
+//Debit us and credit remote
+func (s *Swap) chargeLocal(event *p2p.PeerEvent, amount *big.Int) {
```

Comment:
using big.Int adds so much complexity, can we just use int64 and convert if really needed?

Reply:
I was using `big.Int` just only because the old implementation was doing so, so I just assumed it had been evaluated and decided previously. I have no issue in converting to `int64`

Reply:
pls do, use big int only when needed for ABI contract interface